### PR TITLE
Executor to use any port if not configured. Added REST APIs to activate / deactivate executor

### PR DESCRIPTION
**Pull request by suvodeep-pyne**
When an executor server is launched, it uses the configured port or falls back to a default port. This behavior has been modified such that when the executor is not configured to a specific port it defaults to any available port that it can find on the machine. The advantage is that the user doesn't need to worry about port collisions when he/she doesn't need a specific port.

  Also, The executor now automatically adds an entry about itself to the DB post launch. Also, it exposes the following REST APIs to allow users to activate / deactivate itself
  example:
  http://localhost:1234/executor?action=activate
  http://localhost:1234/executor?action=deactivate

On a successful start, the executor writes the port to a file called executor.port by default.

Contains small refactors like consolidating constants. simplifying function signature.

No automated tests were added due to the nature of the change. No additional functionality was added by existing pieces of code was used. However, the changes were tested manually.

### CODE REVIEW

**Comment by HappyRay** [comment]
"No additional functionality was added by existing pieces of code was used." Do you mean " no other additional functionality was added? "

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java [hunk 1]
```diff
@@ -129,11 +129,13 @@ private DataSourceUtils() {
 
     private static MonitorThread monitorThread = null;
 
+    private final String url;
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: Why this change?

File: azkaban-common/src/main/java/azkaban/server/Constants.java [hunk 2]
```diff
@@ -16,6 +16,11 @@
 
 package azkaban.server;
 
-public class ServerConstants {
+public class Constants {
 	public static final String AZKABAN_SERVLET_CONTEXT_KEY = "azkaban_app";
+	public static final String AZKABAN_PROPERTIES_FILE = "azkaban.properties";
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/server/Constants.java]: Is this class shared between the web server and executor?

File: azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java [hunk 3]
```diff
@@ -134,34 +171,35 @@ public AzkabanExecutorServer(Props props) throws Exception {
...
+    } catch (ExecutorManagerException e) {
+      logger.error("Error inserting executor entry into DB", e);
+      Throwables.propagate(e);
     }
+  }
 
-    logger.info("Azkaban Executor Server started on port " + portNumber);
+  private void dumpPortToFile() {
+    // By default this should write to the working directory
+    String filename = props.getString(AZKABAN_EXECUTOR_PORT_FILE, "executor.port");
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java]: Is it necessary to make the port file name configurable?

File: azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java [hunk 4]
```diff
@@ -134,34 +171,35 @@ public AzkabanExecutorServer(Props props) throws Exception {
...
+    try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
+      writer.write(String.valueOf(getPort()));
+      writer.write("\n");
+    } catch (IOException e) {
+      logger.error(e);
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java]: Our automation may rely on the port file being there. Should we allow the exception to be bubbled up?

File: azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java (hunk 1 above)
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java]: I wanted to log the url in case of DB connection failures. 
Apparently there is no easy API to access the url from that class.

File: azkaban-common/src/main/java/azkaban/server/Constants.java (hunk 2 above)
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/server/Constants.java]: yes. everything in azkaban-common is.

File: azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java (hunk 3 above)
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java]: No. In fact I don't see any advantage in doing it.
I just did it since I didn't see any disadvantage either. It allows people to put whatever name they like.
This can be changed to a constant.

File: azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java (hunk 4 above)
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/AzkabanExecutorServer.java]: yes. I updated the code

**Comment by HappyRay** [comment]
LGTM.

In the future, the API change seems to be independent of the port change. I would suggest making the independent changes separate pull requests.
